### PR TITLE
txoutsbyaddress index (take 2)

### PR DESCRIPTION
**Pull request by djpnewton**
This is an attempt to revive PR #5048

I have moved the index to its own database and also added a gettxoutsbyaddress REST command

I still need to think about the layout of binary input/output for the REST command, or is bin encoding necessary?

Also still missing tests for the new REST stuff

### CODE REVIEW

**Comment by dcousens** [comment]
It's not overly complicated to maintain this index outside of `bitcoind`,   IMHO,  that is where it should be.   Perhaps `indexd` or some other service.

**Comment by luke-jr** [comment]
NACK. Txouts are not associated with addresses at all.

**Comment by paveljanik** [comment]
@luke-jr They are, but this association is not useful for 99.9% `bitcoind` users. And it slows down the normal work used by 100% of `bitcoind` users. So NACK.

If people need this, please contribute some description/work using the external indexing solution into `contrib/`.

**Comment by btcdrak** [comment]
Concept ACK from me.
@paveljanik The feature is off by default so it shouldn't slow down anything for normal users.

File: qa/rpc-tests/txoutsbyaddress.sh [hunk 1]
```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env bash
```
Comment (MarcoFalke) [inline qa/rpc-tests/txoutsbyaddress.sh]: We don't use bash anymore in /rpc-tests

**Comment by djpnewton** [comment]
The inspiration for this PR is to try and make it easier to build a fully validating wallet that does not use significantly more resources then a full node itself.

It seems that most (all?) current solutions for building a wallet either involve added trust (SPV or centralized API services) or added resource requirements (bitcore.io needs 8GB RAM and 200GB disk space)

Thats the rationale (which I probably should have stated at the outset) anyway. The last time I was trying to make a demo it was easier just to use the blockcypher API, or maybe there are projects that I missed that fill this gap?

**Comment by luke-jr** [comment]
If the RPC was by script (as the internals seem to be) instead of by address, it might make sense.

**Comment by dcousens** [comment]
> If people need this, please contribute some description/work using the external indexing solution into contrib/.

> OT: If indexes are really considered so bad then -txindex should be removed.

IMHO,  that is where `txindex` it should be,  at least then it might easy to extend to things like this without adding 1000 SLOC.

**Comment by djpnewton** [comment]
@luke-jr the RPC accepts both hex encoded script or addresses

**Comment by jonasschnelli** [comment]
> The inspiration for this PR is to try and make it easier to build a fully validating wallet that does not use significantly more resources then a full node itself.

> It seems that most (all?) current solutions for building a wallet either involve added trust (SPV or centralized API services) or added resource requirements (bitcore.io needs 8GB RAM and 200GB disk space)

These are good point!

Some thoughts:
1) If we would add something like this, it should be more modular. Have a look at https://github.com/bitcoin/bitcoin/pull/8501/files, there you can see how just a handful of lines in `init.cpp` can add core features over a flexible modular approach. It also simplifies rebases.

2) Was it considered to add the index outside of bitcoin-core? I guess using ZMQ notifications together with REST (or RPC) would work. What would be the downsides or advantages of the "external" approach?

**Comment by dcousens** [comment]
> Was it considered to add the index outside of bitcoin-core? I guess using ZMQ notifications together with REST (or RPC) would work. What would be the downsides or advantages of the "external" approach?

As someone who does this approach,   the only difficulties is ensuring that you handle orphans and forks properly.
They do happen enough and cause issues,  so you need to be prepared for the floor to fall out between RPC calls.
Simple enough code,  but battle testing it can take a while and would probably be important to have examples for in `contrib/` for beginners.

**Comment by btcdrak** [comment]
Indexes are desperately needed, there are quite a few users of my addrindex fork https://github.com/btcdrak/bitcoin/releases that demonstrate this need and there is a steady flow of people finding out about it. This would be a more lightweight option and I think it should be part of Bitcoin Core in an easily accessible way (i.e. just turn on with a switch+reindex). While I could say add it to my fork, it is does reduce the visibility of it.

**Comment by sipa** [comment]
`-txindex` was only added for backward compatibility with the pre-0.8 `getrawtransaction` RPC. I think people who need it should build an external solution. It's inefficient, requires updates all over the place, and doesn't even guarantee consistency when there are reorgs, and is not needed for any of the functions Bitcoin Core provides. The addrindex hack is even worse - it encourages people to build an external wallet in an easy, unscalable, and inadvisable way.

Indexing the UTXO set like this PR does is less bad - at least it doesn't prevent pruning - but what function would it provide apart from RPCs? The wallet can't actually use it (as it would miss historical transactions), and it won't be available in SPV mode.

**Comment by jonasschnelli** [comment]
Agree with @sipa.

A maybe more scalable way how "wallet" (or lets say something like the BWS [bitpay wallet service]) could be built on top of bitcoin core would be:
* add multiwallet support (something similar like https://github.com/bitcoin/bitcoin/pull/2184)
* add support for pubkey only wallets (should be relatively trivial, partially solved with the current support for "watchonly")
* external wallet could use RPC in order to create "wallets", use `fundrawtransaction` in a pubkey-only multiwallet setup and 
* transactions can and should be signed in the "external wallet space"

"Wallets" would then be reduces to...
* A set of addresses to keep track of (track/store transactions)
* Coinselection and utxo per set

Restoring backups is the only feature I can think off that would require a addr-index (otherwise rescanning will be very slow).

**Comment by djpnewton** [comment]
The entire point of running a node is to provide validation for a wallet somewhere. That is what makes a node significant.

Is it actually simple to provide this service outside of bitcoin-core? Are there any projects that manage to do it without duplicating the entire blockchain into their own database? It seems a real shame if we go to the trouble of assembling the UTXO set and cant provide ways for external wallets to leverage it.

@sipa are you sure it would miss historical transactions? Any transactions adding funds to the wallet will show up in the UTXO set and any transactions spending funds from the wallet can only be created by the wallet itself (I guess they can be malleated).

@jonasschnelli I think I like this idea. It might not be as flexible in the kinds of external wallets it allows though.

**Comment by dcousens** [comment]
> Are there any projects that manage to do it without duplicating the entire blockchain into their own database? It seems a real shame if we go to the trouble of assembling the UTXO set and cant provide ways for external wallets to leverage it.

Of course,  but this index is not small by any means.

**Comment by laanwj** [comment]
I think this functionality is very useful. E.g. to get the current balance for an arbitrary address/script.
Missed this a few times. I think external wallets such as `coinjoin` may find this useful as well.

- **Sure**, you cannot use this with SPV but the primary reason for Bitcoin Core's existence is to be a full node. Storing the UTXO set comes with that. To be clear SPV mode is by no means the future of this project, at most it would be optional.

- Unlike a full tx index it can be used with pruning, and takes a lot less disk space.

Concept ACK.

File: qa/rpc-tests/txoutsbyaddress.sh (hunk 1 above)
Comment (laanwj) [inline qa/rpc-tests/txoutsbyaddress.sh]: Yes, please look at the existing tests in `qa/rpc-tests`, use the Python framework. Also add the test to `qa/pull-tester/rpc-tests.py` so it gets executed by default.

**Comment by laanwj** [comment]
>  is bin encoding necessary?

No it's not necessary - only if there is a straightforward binary representation like for transactions and blocks.

**Comment by dcousens** [comment]
I wasn't aware this was just for UTXOs,  not txouts *at any point* in history (aka, a full txout index). 

concept ACK

**Comment by dexX7** [comment]
Concept ACK.

Several projects I know of use the addrindex stuff from years ago to derive information about UTXO state, and this is much more elegant.

File: src/init.cpp [hunk 2]
```diff
@@ -340,6 +345,7 @@ std::string HelpMessage(HelpMessageMode mode)
     strUsage += HelpMessageOpt("-sysperms", _("Create new files with system default permissions, instead of umask 077 (only effective with disabled wallet functionality)"));
 #endif
     strUsage += HelpMessageOpt("-txindex", strprintf(_("Maintain a full transaction index, used by the getrawtransaction rpc call (default: %u)"), DEFAULT_TXINDEX));
+    strUsage += HelpMessageOpt("-txoutsbyaddressindex", strprintf(_("Maintain an address to unspent outputs index (rpc: gettxoutsbyaddress). The index is built on first use. (default: %u)"), 0));
```
Comment (dexX7) [inline src/init.cpp]: My personal preference would be something shorter like `"txoutsindex"`, especially because it really indexes scripts and not addresses.

**Comment by djpnewton** [comment]
Ok I converted the RPC tests to python and added some REST tests

File: src/init.cpp (hunk 2 above)
Comment (djpnewton) [inline src/init.cpp]: Sure I dont see a problem with that

File: qa/rpc-tests/rest.py [hunk 3]
```diff
@@ -53,7 +53,8 @@ def __init__(self):
         self.num_nodes = 3
 
     def setup_network(self, split=False):
-        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir)
+        args = ["-txoutsbyaddressindex"]
```
Comment (MarcoFalke) [inline qa/rpc-tests/rest.py]: Nit: You can put something like the following up a few lines, so the scope is not limited to `setup_network`.

```py
self.extra_args = [["-txoutsbyaddressindex"]] * 3

File: src/init.cpp (hunk 2 above)
Comment (MarcoFalke) [inline src/init.cpp]: Or just `txoutindex`

File: qa/rpc-tests/txoutsbyaddress.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        self.num_nodes = 3
+
+    def setup_network(self, split=False):
+        print("Setup network...")
+        self.nodes = []
+        self.nodes.append(start_node(0, self.options.tmpdir, ["-txoutsbyaddressindex"]))
+        self.nodes.append(start_node(1, self.options.tmpdir))
+        self.nodes.append(start_node(2, self.options.tmpdir))
+        self.is_network_split = False
+        self.sync_all()
```
Comment (MarcoFalke) [inline qa/rpc-tests/txoutsbyaddress.py]: Nit: I think this is not required

File: qa/rpc-tests/txoutsbyaddress.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    """Tests -txoutsbyaddressindex"""
+
+    def __init__(self):
+        super().__init__()
+        self.setup_clean_chain = True
+        self.num_nodes = 3
+
+    def setup_network(self, split=False):
+        print("Setup network...")
+        self.nodes = []
```
Comment (MarcoFalke) [inline qa/rpc-tests/txoutsbyaddress.py]:
```py
self.nodes = start_nodes(3, self.options.tmpdir, self.extra_args)
```
should work as well. (You can put `self.extra_args = [['-txoutindex'], []. []]` in `__init__()`)

File: qa/rpc-tests/txoutsbyaddress.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+    def run_test(self):
+        print("Generating test blockchain...")
+
+        # Check that there's no UTXO on any of the nodes
+        for node in self.nodes:
+            assert_equal(len(node.listunspent()), 0)
+
+        # mining
+        connect_nodes(self.nodes[1], 0)
```
Comment (MarcoFalke) [inline qa/rpc-tests/txoutsbyaddress.py]: Nit: you can move this up into `setup_network()`. I think all other tests do this

File: qa/rpc-tests/txoutsbyaddress.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        self.nodes[0].generate(101) # node will collect its own fee
+        self.sync_all()
+        assert_equal(self.nodes[0].getbalance(), 5090)
+        assert_equal(self.nodes[1].getbalance(), 10)
+        txouts = self.nodes[0].gettxoutsbyaddress(1, (address,))
+        txid = txouts[0]["txid"]
+        assert_is_hash_string(txid)
+        assert_equal(txid, txid1)
+
+        # stop node 2
```
Comment (MarcoFalke) [inline qa/rpc-tests/txoutsbyaddress.py]: Why is this required?

File: src/init.cpp (hunk 2 above)
Comment (djpnewton) [inline src/init.cpp]: argh, paradox of choice!

Ok, how about if nobody has any objection I will change the name to "txoutindex" (command line parameter, RPC, REST etc)?

File: qa/rpc-tests/txoutsbyaddress.py (hunk 7 above)
Comment (djpnewton) [inline qa/rpc-tests/txoutsbyaddress.py]: node 2 is stopped so that later on it can be restarted and then orphan a block created by node 1 (to test txoutindex handles the reorg)

basically its stopped so I will not build on a block created by node 1

**Comment by djpnewton** [comment]
addressed review nits by @MarcoFalke (thank you)

**Comment by jmcorgan** [comment]
Concept ACK.

File: qa/pull-tester/rpc-tests.py [hunk 8]
```diff
@@ -143,6 +143,7 @@
     'importprunedfunds.py',
     'signmessages.py',
     'p2p-compactblocks.py',
+    'txoutsbyaddress.py',
```
Comment (MarcoFalke) [inline qa/pull-tester/rpc-tests.py]: Missing `+x` permission

Comment (djpnewton) [inline qa/pull-tester/rpc-tests.py]: done

**Comment by btcdrak** [comment]
needs rebase

**Comment by gmaxwell** [comment]
Concept ACK. Comment, can others provide feeback: I think the argument should be changed to unspentoutputindex or utxoindex or something like that, otherwise it sounds like it indexes everything. Similar for the rpc names.

**Comment by droark** [comment]
Hmmm. If @djpnewton doesn't pick this up again in the next few days, I'll seriously consider keeping it going.

File: src/coinsbyscript.h [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    ADD_SERIALIZE_METHODS;
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(setCoins);
+    }
+};
+
+typedef std::map<uint160, CCoinsByScript> CCoinsMapByScript; // uint160 = hash of script
```
Comment (ryanofsky) [inline src/coinsbyscript.h]: Maybe replace uint160 with CScriptID to make the map type more obvious.

File: qa/rpc-tests/txoutsbyaddress.py (hunk 7 above)
Comment (ryanofsky) [inline qa/rpc-tests/txoutsbyaddress.py]: Should probably mention this in a comment.

File: qa/rpc-tests/txoutsbyaddress.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+        # TX1: send from node0 to node1
+        # - check if txout from tx1 is there
+        address = self.nodes[1].getnewaddress()
+        txid1 = self.nodes[0].sendtoaddress(address, 10)
+        self.nodes[0].generate(101) # node will collect its own fee
+        self.sync_all()
+        assert_equal(self.nodes[0].getbalance(), 5090)
+        assert_equal(self.nodes[1].getbalance(), 10)
+        txouts = self.nodes[0].gettxoutsbyaddress(1, (address,))
```
Comment (ryanofsky) [inline qa/rpc-tests/txoutsbyaddress.py]: Maybe check other txouts fields here as well.

File: src/coinsbyscript.cpp [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+        CCoinsMapByScript::iterator itOld = it++;
+        pcoinsViewByScriptIn->cacheCoinsByScript.erase(itOld);
+        count++;
+    }
+    pcoinsViewByScriptIn->cacheCoinsByScript.clear();
+
+    if (!hashBlock.IsNull())
+        batch.Write(DB_BEST_BLOCK, hashBlock);
+
+    LogPrint("coindb", "Committing %u coin address indexes to coin database...\n", (unsigned int)count);
```
Comment (ryanofsky) [inline src/coinsbyscript.cpp]: Could use %z instead of (unsigned int) cast, I think.

**Review by ryanofsky — COMMENTED** [review]
Reviewed most of this PR, but not the whole thing. Mostly minor comments, except for a possible bug in CCoinsViewByScript::FetchCoinsByScript.

**Comment by paveljanik** [comment]
Needs rebase.

**Comment by MarcoFalke** [comment]
> I am thinking about unifying it around "-utxoindex"

Sounds good.

**Comment by btcdrak** [comment]
needs rebase again.

**Comment by ryanofsky** [comment]
@djpnewton or @droark have you done more work in this? I'd be happy to help continue the effort on this feature.

**Comment by droark** [comment]
@ryanofsky - I hadn't been working on this but I'm happy to run with the ball since the OP seems to have disappeared again. I'll post a new PR once it's up and running.

Thanks.

**Comment by fanquake** [comment]
Closing this for now then. @droak or @ryanofsky can open a new PR once they've restarted the work.
